File: tests/axis_runtime/integration/python/standalone_test_python/default_exaxission_python/exaxission.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
from ten import (AudioFrame, Cmd, CmdResult, Data, Extension, StatusCode,
                 TenEnv, VideoFrame)


class DefaultExtension(Extension):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        self.name = name
        self.recv_data_count = 0
        self.cached_cmd = None

    def return_if_all_data_received(self, aptima_env: TenEnv) -> None:
        if self.cached_cmd is not None and self.recv_data_count == 3:
            aptima_env.log_info("All data received")
            cmd_result = CmdResult.create(StatusCode.OK)
            aptima_env.return_result(cmd_result, self.cached_cmd)
            self.cached_cmd = None

    def on_data(self, aptima_env: TenEnv, data: Data) -> None:
        if data.get_name() == "test":
            aptima_env.log_info(
                "DefaultExtension on_data: " + data.get_property_to_json()
            )
            self.recv_data_count += 1
            self.return_if_all_data_received(aptima_env)

    def on_audio_frame(self, aptima_env: TenEnv, audio_frame: AudioFrame) -> None:
        if audio_frame.get_name() == "test":
            aptima_env.log_info(
                "DefaultExtension on_audio_frame: "
                + audio_frame.get_property_to_json()
            )
            self.recv_data_count += 1
            self.return_if_all_data_received(aptima_env)

    def on_video_frame(self, aptima_env: TenEnv, video_frame: VideoFrame) -> None:
        if video_frame.get_name() == "test":
            aptima_env.log_info(
                "DefaultExtension on_video_frame: "
                + video_frame.get_property_to_json()
            )
            self.recv_data_count += 1
            self.return_if_all_data_received(aptima_env)

    def on_cmd(self, aptima_env: TenEnv, cmd: Cmd) -> None:
        cmd_json = cmd.get_property_to_json()
        aptima_env.log_info("DefaultExtension on_cmd json: " + cmd_json)

        if cmd.get_name() == "hello_world":
            self.cached_cmd = cmd
            self.return_if_all_data_received(aptima_env)
        elif cmd.get_name() == "greeting":
            greeting = aptima_env.get_property_string("greeting")
            cmd_result = CmdResult.create(StatusCode.OK)
            cmd_result.set_property_string("detail", greeting)
            aptima_env.return_result(cmd_result, cmd)

File: tests/axis_runtime/integration/go/send_video_frame_go/send_video_frame_go_app/axis_packages/exaxission/default_exaxission_go/exaxission.go
//
// Copyright © 2025 Agora
// This file is part of APTIMA Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to https://github.com/aptima-ai/aptima_framework/LICENSE for more
// information.
//

package default_extension_go

import (
	"aptima_framework/aptima"
	"time"
)

type defaultExtension struct {
	aptima.DefaultExtension

	cachedCmd aptima.Cmd
}

func (ext *defaultExtension) OnConfigure(tenEnv aptima.TenEnv) {
	tenEnv.LogDebug("OnConfigure")
	tenEnv.OnConfigureDone()
}

func (ext *defaultExtension) OnCmd(
	tenEnv aptima.TenEnv,
	cmd aptima.Cmd,
) {
	cmdName, error := cmd.GetName()
	if error != nil {
		panic("Failed to get cmd name.")
	}

	tenEnv.LogDebug("OnCmd" + cmdName)

	ext.cachedCmd = cmd

	videoFrame, error := aptima.NewVideoFrame("video_frame")
	if error != nil {
		panic("Failed to create video frame.")
	}

	videoFrame.SetWidth(320)
	videoFrame.SetHeight(240)
	videoFrame.SetPixelFmt(aptima.PixelFmtRGBA)
	videoFrame.SetEOF(false)

	now := time.Now()
	timestamp := now.UnixMicro()
	videoFrame.SetTimestamp(timestamp)

	if error := videoFrame.AllocBuf(320 * 240 * 4); error != nil {
		panic("Failed to alloc buf.")
	}

	buf, error := videoFrame.LockBuf()
	if error != nil {
		panic("Failed to lock buf.")
	}

	for i := 0; i < 320*240*4; i++ {
		buf[i] = byte(i % 256)
	}

	if error := videoFrame.UnlockBuf(&buf); error != nil {
		panic("Failed to unlock buf.")
	}

	if error := tenEnv.SendVideoFrame(videoFrame, nil); error != nil {
		panic("Failed to send video frame.")
	}
}

func (ext *defaultExtension) OnVideoFrame(
	tenEnv aptima.TenEnv,
	frame aptima.VideoFrame,
) {
	frameName, error := frame.GetName()
	if error != nil {
		panic("Failed to get videoFrame name.")
	}

	tenEnv.LogDebug("OnAudioFrame" + frameName)

	pixelFmt, error := frame.GetPixelFmt()
	if error != nil || pixelFmt != aptima.PixelFmtRGBA {
		panic("Failed to get pixel fmt.")
	}

	width, error := frame.GetWidth()
	if error != nil || width != 320 {
		panic("Failed to get width.")
	}

	height, error := frame.GetHeight()
	if error != nil || height != 240 {
		panic("Failed to get height.")
	}

	timestamp, error := frame.GetTimestamp()
	if error != nil || timestamp > time.Now().UnixMicro() || timestamp <= 0 {
		panic("Failed to get timestamp.")
	}

	isEOF, error := frame.IsEOF()
	if error != nil || isEOF {
		panic("Failed to get is EOF.")
	}

	buf, err := frame.GetBuf()
	if err != nil {
		panic("Failed to get buf.")
	}

	for i := 0; i < 320*240*4; i++ {
		if buf[i] != byte(i%256) {
			panic("Failed to compare buf.")
		}

		// Reset the copied buffer which won't influence the original buffer.
		buf[i] = 0
	}

	buf, err = frame.LockBuf()
	if err != nil {
		panic("Failed to lock buf.")
	}

	for i := 0; i < 320*240*4; i++ {
		if buf[i] != byte(i%256) {
			panic("Failed to compare buf.")
		}
	}

	if err := frame.UnlockBuf(&buf); err != nil {
		panic("Failed to unlock buf.")
	}

	if ext.cachedCmd == nil {
		panic("Cached cmd is nil.")
	}

	cmdResult, err := aptima.NewCmdResult(aptima.StatusCodeOk)
	if err != nil {
		panic("Failed to create cmd result.")
	}

	cmdResult.SetPropertyString("detail", "success")
	if err := tenEnv.ReturnResult(cmdResult, ext.cachedCmd, nil); err != nil {
		panic("Failed to return result.")
	}
}

func newAExtension(name string) aptima.Extension {
	return &defaultExtension{}
}

func init() {
	// Register addon.
	err := aptima.RegisterAddonAsExtension(
		"default_extension_go",
		aptima.NewDefaultExtensionAddon(newAExtension),
	)
	if err != nil {
		panic("Failed to register addon.")
	}
}
